Realtime_zshim shim CLI

po09i opened the pull request:
## Description
This PR adds the ability to do realtime zshimming with the CLI. The code is based on the realtime_zshim script Matlab repo.

Some definitions for the gradients: 
- Gslice, Gphase, Gfreq are in the patient coordinate system.
- Gx, Gy, Gz are in the scanner coordinate system

Some background about z-shimming workflow:
- Acquire a multi-echo fieldmap data (eg could be sagittal)
- B0 fieldmap is calculated for each voxel
- Use `phys_gradient` to calculate voxelwise Gx, Gy, Gz
- calculate RIRO, STATIC for Gx, Gy and Gz
- Resample and regrid to the axial scan (So we have what would be Gx, Gy and Gz mapped on the axial scan voxels)
- Calculate Gslice of the axial scan  using `phys_to_vox_gradient`
- We call it "z-shimming", where the "z" is relative to the coordinate system of the image (ie: the slice select), not the physical coordinate system. Ie: there will be contribution from the X and Y gradients as well for non "pure" axial orientation which is why we want Gslice.

## Linked issues
Related to #101
Fixes to #37

jcohenadad commented:
Working prototype on b6f246a2c240cda3a3c53c5faead92256fbe0af4

To build Matlab doc:
```matlab
FileList = genToolboxHelp('/Users/julien/code/shimming-toolbox', 'shimming-toolbox', {'Ui', 'tests', 'misc', 'Img', 'external', 'example', 'docs', 'doc', 'Coils', 'aux_hardware'}, true, false)
```

Produces:
![image](https://user-images.githubusercontent.com/2482071/85066257-97427f00-b17c-11ea-918e-14964d3cb8e2.png)

jcohenadad commented:
Adding a Contents.m file within each folder enables to do this:
~~~
>> help imutils
  
 
  Files
    dicom_to_nifti      - 
    get_nii_coordinates - 
    load_niftis         - 
    read_nii            - Load NIfTI image and header; and, when present, the accompanying .json sidecar
~~~

jcohenadad commented:
Ready to be tested: 
- Checkout this branch (commit bde1642d53d1faa35fba1efff8885b1b0bb0e288)
- Generate the doc: 
```matlab
startup
FileList = genToolboxHelp('/Users/julien/code/shimming-toolbox', 'shimming-toolbox', {'doc', 'docs'}, false, true)
```
- Open the doc:
```matlab
doc
```

The following window should pop up:
![Screen Shot 2020-06-19 at 12 09 37 PM](https://user-images.githubusercontent.com/2482071/85154855-01fbc500-b226-11ea-8115-b2f724784d95.png)

po09i commented:
I followed the procedure and had the problem with `builddocsearchdb` when running `genToolboxHelp`. I had to run restart Matlab and run `genToolboxHelp` again to fix the problem.

We probably want to create a template on how to create new functions so that both the website and the matlab documentation look good.

Every header page shows a page not found :
![Screen Shot 2020-06-19 at 14 01 42](https://user-images.githubusercontent.com/47249340/85166531-6a9e6e00-b235-11ea-9915-b73ddb8f041e.png)

jcohenadad commented:
> We probably want to create a template on how to create new functions so that both the website and the matlab documentation look good.
yup! this is the purpose of https://github.com/shimming-toolbox/shimming-toolbox/issues/133. Out of the scope of this PR. Once merged, we can work on this issue.

> Every header page shows a page not found :
yup! i'll find a fix to it before merging.

jcohenadad commented:
> I followed the procedure and had the problem with builddocsearchdb when running genToolboxHelp. I had to run restart Matlab and run genToolboxHelp again to fix the problem.

🤔  mmh that's annoying. if you could find a better fix to it that would be awesome

rtopfer commented:
> EDIT: this mush be a new Matlab feature, but even without Contents.m file for non-packages, it does list all the files. So let's get rid of this Contents.m file.

Actually it's an old feature that, indeed, doesn't seem completely worked out: It's nice that it prints the contents automatically without the Contents.m file, but it seems to me having a package overview is sensible, and since `help package_name` displays Contents.m, this seems like the place to put it? E.g. for all the shim material I had started [this WIP](https://github.com/shimming-toolbox/shimming-toolbox/blob/rt/issue75-ShimSpecs-json/%2Bb0shim/Contents.m) for issue #75

jcohenadad commented:
> Actually it's an old feature that, indeed, doesn't seem completely worked out: It's nice that it prints the contents automatically without the Contents.m file, but it seems to me having a package overview is sensible, and since help package_name displays Contents.m, this seems like the place to put it? E.g. for all the shim material I had started this WIP for issue #75

funny you are writing this now, as I am just realizing that the issue described [here](https://github.com/shimming-toolbox/shimming-toolbox/pull/134#issuecomment-646799509) is caused by a missing Contents.m file 😬. So I am currently creating one per folder, with brief description about the purpose of each folder/package.

jcohenadad commented:
Missing page issue fixed in c240021e064aa64413e7787cd5940ebdb30a78e5. Output looks like this:
![Screen Shot 2020-06-19 at 3 11 56 PM](https://user-images.githubusercontent.com/2482071/85172560-58293200-b23f-11ea-841e-a8d457e3f847.png)

I suggest to indicate with (package) when a folder is a package.

jcohenadad commented:
even better:
![Screen Shot 2020-06-19 at 3 19 23 PM](https://user-images.githubusercontent.com/2482071/85173090-6f1c5400-b240-11ea-9265-71e6cfb2c45c.png)

po09i commented:
> Missing page issue fixed in [c240021](https://github.com/shimming-toolbox/shimming-toolbox/commit/c240021e064aa64413e7787cd5940ebdb30a78e5).

The missing page issue is also for all the other headers, not only +imutils as shown for the "coils" section here :

https://github.com/shimming-toolbox/shimming-toolbox/pull/134#issuecomment-646799509

jcohenadad commented:
> > Missing page issue fixed in [c240021](https://github.com/shimming-toolbox/shimming-toolbox/commit/c240021e064aa64413e7787cd5940ebdb30a78e5).
> 
> The missing page issue is also for all the other headers, not only +imutils as shown for the "coils" section here :
> 
> [#134 (comment)](https://github.com/shimming-toolbox/shimming-toolbox/pull/134#issuecomment-646799509)

yes, i was just looking for a proper solution for imutils, which can then be applied to the other folders. At this point, given the large refactoring ahead, i suggest we don't spend time populatin entries that will disappear in the short term. The purpose here is to settle on a set of conventions, that will then be applied to the code at large (in subsequent PRs).

jcohenadad commented:
> Current problem is building the doc the first time throws an error with `builddocsearchdb` will open an issue for it.

https://github.com/shimming-toolbox/shimming-toolbox/issues/144

po09i committed:
Initial statis shim set up

jcohenadad committed:
Merge branch 'master' into ad/sequencer-realtime-zshim

jcohenadad committed:
Moved realtime_zshim to CLI

jcohenadad committed:
Added test for realtime_zshim

po09i committed:
Completed test set up

po09i committed:
Added to set up as a CLI

po09i committed:
Merged master into ad/sequencer-realtime-zshim

po09i commented:
Read the docs timed out but when looking at the log it says it succeeded. I think this is a readthedocs bug not one on our side. We will see on the next push.

po09i committed:
Reduce shimming volume

po09i committed:
Add plot of currents through time

po09i committed:
Adds ability to see output error messages from runner.invoke

po09i commented:
There is a bug in lsq_optimizer which is why the build is failing. I opened an issue #152

jcohenadad committed:
Merge branch 'master' into ad/sequencer-realtime-zshim

jcohenadad committed:
Changed variable name

jcohenadad committed:
Clarified comments, added TODO, minor refactoring

- Added bounds to sequential_zslice() because of bug: https://github.com/shimming-toolbox/shimming-toolbox/issues/152
- Commented out Click decorator because not working in Pycharm: https://github.com/shimming-toolbox/shimming-toolbox/issues/156

jcohenadad committed:
Added TODO

po09i committed:
Change debug paths and plot correct currents

po09i committed:
Sanity check for z component

po09i committed:
Added TODO for pmu_data, Related to #149

po09i committed:
Merge branch 'master' into ad/sequencer-realtime-zshim

po09i committed:
Added pmu timestamps and pressures

po09i committed:
Add resp path, json path and fix fieldmap path

po09i committed:
Udated click options and commented debug to verify test

jcohenadad committed:
Added TODOs

po09i committed:
Add fit for static and riro linear regression between pressure and field

po09i committed:
Added plots and refactored other plots

po09i committed:
Add comments

po09i committed:
Rescale PMU and correct index for z component

po09i committed:
Add ability to mask with threshold

po09i committed:
Add resampling of fieldmap, riro and static to target acquisition

po09i committed:
Use nibabel instead of nilearn for resampling

po09i committed:
Added gradient calculation

po09i committed:
Output pmu from 0 to 4095 to be consistent with matlab code

po09i committed:
Calculate mean for each slice

po09i committed:
Write correction coefficients to a file

po09i committed:
Comments

po09i committed:
Add mean pressure in calculation to account if bellow moves

po09i committed:
Add local full acquisition and refine cube mask

po09i committed:
Add plot of evolution of riro and static coefficients

jcohenadad committed:
Reorganized code, added comments

jcohenadad committed:
Now outputting Gz instead of fieldmap

jcohenadad committed:
Added DEBUG global var

jcohenadad committed:
Added fig_ prefix to all output figures

jcohenadad committed:
Added progress bar to fitting

Progress bar not showing though-- will need to fix that

po09i committed:
Apply mask on anat instead of the fieldmap

po09i committed:
Commented out unecessary code

po09i committed:
Improve masking and mean calculation

po09i committed:
add outpout option to output figures and text file

po09i committed:
Clean up and added more to DEBUG

po09i committed:
Removed plot that only works for fieldmap fitting

po09i committed:
Add isclose to input masks with sligth affine variation

po09i committed:
Add meshgrid and local mask for validation

po09i committed:
Merge branch 'master' into ad/sequencer-realtime-zshim

po09i committed:
Update gradient to use meshgrid points

po09i committed:
Add scaling by RMS to RIRO to have comparable figures across sunjects

po09i committed:
Update figures

evaalonsoortiz commented:
In the matlab code, we are doing option 2, as listed above (see [here](https://github.com/shimming-toolbox/shimming-toolbox-matlab/blob/7f0ba2d480a697a16a0d4b20154c61171f7e13ca/Img/MaRdI.m#L1150)). I think both options are equivalent, but with option 1 you are limited by the fact that you may have a small number of slices for the MGRE scan, which would introduce errors in the Gslice calculation.

po09i commented:
I think the relevant line is [here](https://github.com/shimming-toolbox/shimming-toolbox-matlab/blob/7f0ba2d480a697a16a0d4b20154c61171f7e13ca/example/realtime_zshim.m#L191). From what I understand, this gives a gradient which is wrong if Gslice and Gz are not aligned. The gradient attempts to give Gslice but if the image is rotated, the z voxel positions given are right but are not appropriate to calculate the gradient. The coordinates should be along Gslice and not along Gz. I'll reiterate that I think the current implementation works for past experiments since the fieldmap is aligned with the scanner coordinate system.

I think the `resliceImg()` [function](https://github.com/shimming-toolbox/shimming-toolbox-matlab/blob/7f0ba2d480a697a16a0d4b20154c61171f7e13ca/Img/MaRdI.m#L1150) you are mentioning remaps the gradient to the appropriate coordinates of the axial scan but if the initial fieldmap gradient is wrong (due to the above explanation) or if the gradient is along Gslice (Gslice of the fieldmap where Gslice is not along Gz), then the resliced img will be wrong as well.

po09i committed:
Remove comments related to sequencer. Couls be relevant for #101

po09i committed:
Rafactor cli into an API and a CLI

po09i committed:
Started comments

po09i committed:
Merge branch 'master' into ad/sequencer-realtime-zshim

po09i committed:
Add wrapper to resample_from_to to allow 2D and 4D input data

po09i committed:
Refactor to use shimming toolbox gradient and resample functions

po09i committed:
Add support for all acquisition orientations. The fieldmap and the anat image no longer have to line up

po09i committed:
Add comments and changed pmu parameter of API

po09i committed:
Fix bugs, remove debug, and update comments

po09i committed:
Add tests for realtime_zshim API

po09i committed:
Clean up

po09i committed:
More clean up

jcohenadad commented on shimmingtoolbox/coils/coordinates.py:
```diff
@@ -125,3 +127,49 @@ def phys_to_vox_gradient(gx, gy, gz, affine):
+
+def resample_from_to(nii_from_img, nii_to_vox_map, order=3, mode='constant', cval=0., out_class=nib.Nifti1Image):
+    """ Wrapper to nibabel's ``resample_from_to`` function. Resample image `from_img` to mapped voxel space
+    `to_vox_map`. The wrapper adds support for 2D input data (adds a singleton) and 4D timeseries support.
```
```suggestion
    `to_vox_map`. The wrapper adds support for 2D input data (adds a singleton) and for 4D time series.
```

jcohenadad commented on shimmingtoolbox/shim/realtime_zshim.py:
```diff
@@ -0,0 +1,259 @@
+def realtime_zshim(nii_fieldmap, nii_anat, pmu, json_fmap, nii_mask_anat=None, path_output=None):
+    """ This function will generate static and dynamic (due to respiration) Gz components based on a fieldmap time
+    series and respiratory trace information obtained from Siemens bellows  (PMUresp_signal.resp). An additional
+    multi-gradient echo (MGRE) magnitiude image is used to generate an ROI and resample the static and dynamic Gz
```
```suggestion
    multi-gradient echo (MGRE) magnitude image is used to generate an ROI and resample the static and dynamic Gz
```

jcohenadad commented on shimmingtoolbox/shim/realtime_zshim.py:
```diff
@@ -0,0 +1,259 @@
+    for i_t in range(nt):
+        masked_fieldmaps[..., i_t] = mask_fmap * fieldmap[..., i_t]
+
+    # Calculate gz gradient
```
```suggestion
    # Calculate Gx, Gy and Gz gradients (in the physical coordinate system)
```

jcohenadad commented on shimmingtoolbox/coils/coordinates.py:
```diff
@@ -125,3 +127,49 @@ def phys_to_vox_gradient(gx, gy, gz, affine):
     return gx_vox, gy_vox, gz_vox
+
+
+def resample_from_to(nii_from_img, nii_to_vox_map, order=3, mode='constant', cval=0., out_class=nib.Nifti1Image):
```
maybe change default order to 2 (trilinear), because 3 creates oscillations (maybe problematic for our applications?) and is longer

jcohenadad commented on shimmingtoolbox/shim/realtime_zshim.py:
```diff
@@ -0,0 +1,259 @@
+
+    # Resample masked_fieldmaps to target anatomical image
+    nii_masked_fieldmaps = nib.Nifti1Image(masked_fieldmaps, nii_fieldmap.affine)
+    nii_resampled_fmap = resample_from_to(nii_masked_fieldmaps, nii_anat, order=2, mode='nearest')
```
order not needed if defaulted to 2

jcohenadad commented on shimmingtoolbox/shim/realtime_zshim.py:
```diff
@@ -0,0 +1,259 @@
+    if is_outputting_figures:
+        nib.save(nii_resampled_static_masked, os.path.join(path_output, 'fig_resampled_static.nii.gz'))
+
+    # Resample riro to target anatomical image
```
high duplication with previous paragraph-- could potentially be a private local function

po09i committed:
Apply suggestions from code review

Co-authored-by: Julien Cohen-Adad <jcohen@polymtl.ca>

po09i committed:
More suggestions from code review

po09i commented:
@jcohenadad
Current approach (Option 1):
- Calculate Gx, Gy, Gz
- Calculate riro and static for each gradient
- Resample to the target anat image
- Calculate Gss

As discussed in the internal meeting, I implemented the approach (Option 2) where we:
- Calculate Gx, Gy, Gz 
- Resample to the target anat image
- Calculate Gss
- Calculate riro and static

Both give the same results as we would expect however, 
Option 1: 8s
Option 2: 52s, neater code

The time difference seems to be related to having to resample the fieldmap to the anat image for all timepoints and axis(x, y and z) instead of doing it only for the static and riro. Since both give the same results, I prefer Option 1.

po09i committed:
Merge branch 'master' into ad/sequencer-realtime-zshim

po09i committed:
Remove json input since it will be in the output of prepare_fieldmap

po09i committed:
Fix values of test

po09i committed:
Merge with master

po09i committed:
Change output folder and check it is not empty

jcohenadad reviewed (approved):
Awesome contribution @po09i 🏅 !!!
